### PR TITLE
samples: cellular: udp: cleanup and add way to shutdown modem

### PR DESCRIPTION
**Pull request by MirkoCovizzi**
This commit adds a new Kconfig option for configuring
the number of data transmissions to the server.
Shutdown the modem library when there are no more
transmissions to execute.
Moved sample to `CI-iot-samples-test` plan.

### CODE REVIEW

File: samples/cellular/udp/src/main.c [hunk 1]
```diff
@@ -126,6 +128,19 @@ static void lte_handler(const struct lte_lc_evt *const evt)
...
+	char resp[32];
+
+	if (nrf_modem_at_cmd(resp, sizeof(resp), "AT+CGMM") == 0) {
+		if (strstr(resp, "nRF9160") != NULL) {
+			return true;
+		}
+	}
+
+	return false;
+}
```
Comment (lemrey) [inline samples/cellular/udp/src/main.c:142]: This sample only works on nRF 9 series SoCs. It may have had support for `qemu_cortex_m3` earlier, but that is no longer the case. I do not see why this function is necessary anymore, can we remove it?

File: samples/cellular/udp/Kconfig [hunk 2]
```diff
@@ -14,6 +14,14 @@ config UDP_DATA_UPLOAD_FREQUENCY_SECONDS
 	int "How often data is transmitted to the server"
 	default 900
 
+config UDP_DATA_UPLOAD_FOREVER
+	bool "Enable that data is transmitted to the server forever"
+	default y
+
+config UDP_DATA_UPLOAD_ITERATIONS
+	int "NUmber of data transmissions to the server before shutdown" if !UDP_DATA_UPLOAD_FOREVER
+	default 5
```
Comment (lemrey) [inline samples/cellular/udp/Kconfig:23]: We shouldn't need two entries for this; one config for the number of iterations should be sufficient.
If set to `-1` let the loop run forever.

File: samples/cellular/udp/src/main.c [hunk 3]
```diff
@@ -239,5 +269,14 @@ int main(void)
 
 	k_work_schedule(&socket_transmission_work, K_NO_WAIT);
 
+#if defined(CONFIG_NRF_MODEM_LIB) && !defined(CONFIG_UDP_DATA_UPLOAD_FOREVER)
```
Comment (lemrey) [inline samples/cellular/udp/src/main.c:272]: You can remove all the occurrences where we check for `NRF_MODEM_LIB`. This sample runs only on nRF9 series boards, so this is always defined. Keep it a separate commit if you can.

File: samples/cellular/udp/src/main.c [hunk 4]
```diff
@@ -174,6 +189,21 @@ static int modem_connect(void)
...
+	int err;
+
+	err = nrf_modem_lib_shutdown();
+	if (err) {
+		printk("Failed to shutdown modem library, error: %d\n", err);
+		return err;
+	}
+
+	return 0;
+}
```
Comment (lemrey) [inline samples/cellular/udp/src/main.c:205]: There is no point to defined this function to call `nrf_modem_lib_shutdown()`; just call `nrf_modem_lib_shutdown()` directly.

File: samples/cellular/udp/src/main.c (hunk 1 above)
Comment (MirkoCovizzi) [reply samples/cellular/udp/src/main.c:142]: I'll update the sample accordingly then. I was under the assumption that qemu target was still required.

File: samples/cellular/udp/Kconfig (hunk 2 above)
Comment (MirkoCovizzi) [reply samples/cellular/udp/Kconfig:23]: I thought about the -1 idea before, but I don't like the fact that a different value has a completely different behavior. While these two options make sense just by reading their name, in the -1 case the user will have to find such behavior in the documentation instead.

Comment (MirkoCovizzi) [reply samples/cellular/udp/Kconfig]: @MechanicV1 made me aware that the `download_client` uses `-1` already in a similar way https://github.com/nrfconnect/sdk-nrf/blob/8f8c2689151063428e0f52645762ebedaecf78ab/subsys/net/lib/download_client/Kconfig#L112 so I updated this sample to align to that.

File: samples/cellular/udp/Kconfig [hunk 5]
```diff
@@ -14,6 +14,13 @@ config UDP_DATA_UPLOAD_FREQUENCY_SECONDS
 	int "How often data is transmitted to the server"
 	default 900
 
+config UDP_DATA_UPLOAD_ITERATIONS
+	int "NUmber of data transmissions to the server before shutdown"
```
Comment (eivindj-nordic) [inline samples/cellular/udp/Kconfig:18]: nit:
```suggestion
	int "Number of data transmissions to the server before shutdown"
```

File: samples/cellular/udp/Kconfig [hunk 6]
```diff
@@ -14,6 +14,13 @@ config UDP_DATA_UPLOAD_FREQUENCY_SECONDS
...
+	default -1
+	help
+	  NUmber of data transmissions to the server before shutdown.
```
Comment (eivindj-nordic) [inline samples/cellular/udp/Kconfig:21]:
```suggestion
	  Number of data transmissions to the server before shutdown.
```

**Review by eivindj-nordic — APPROVED** [review]

File: samples/cellular/udp/Kconfig [hunk 7]
```diff
@@ -14,6 +14,13 @@ config UDP_DATA_UPLOAD_FREQUENCY_SECONDS
 	int "How often data is transmitted to the server"
 	default 900
 
+config UDP_DATA_UPLOAD_ITERATIONS
```
Comment (eivindj-nordic) [inline samples/cellular/udp/Kconfig:17]: Do we want a range here as well, as in the download client?

Comment (MirkoCovizzi) [reply samples/cellular/udp/Kconfig:17]: I thought about it and to me it doesn't make sense to add it

File: samples/cellular/udp/src/main.c [hunk 8]
```diff
@@ -63,6 +65,17 @@ static void socket_transmission_work_fn(struct k_work *work)
...
 
+	if (data_upload_iterations > 0) {
+		data_upload_iterations--;
+		goto transmit;
+	} else if (data_upload_iterations == 0) {
+		k_sem_give(&modem_shutdown_sem);
+		return;
+	} else if (data_upload_iterations < 0) {
+		goto transmit;
+	}
```
[the remark this answers is not in the document]
Comment (MirkoCovizzi) [reply samples/cellular/udp/src/main.c:76]: That looks smart but smart is not always better. I'd rather be explicit and clear. It's not going to make it more efficient anyway.

**Review by MechanicV1 — APPROVED** [review]

File: samples/cellular/udp/src/main.c [hunk 9]
```diff
@@ -63,6 +65,21 @@ static void socket_transmission_work_fn(struct k_work *work)
...
+		goto transmit;
+	} else if (data_upload_iterations == 0) {
+		k_sem_give(&modem_shutdown_sem);
+		return;
+	}
+
+	/* Transmit indefinitely. */
+	if (CONFIG_UDP_DATA_UPLOAD_ITERATIONS == -1) {
+		goto transmit;
+	}
```
Comment (eivindj-nordic) [inline samples/cellular/udp/src/main.c:80]: This is not necessary

File: samples/cellular/udp/Kconfig [hunk 10]
```diff
@@ -14,6 +14,14 @@ config UDP_DATA_UPLOAD_FREQUENCY_SECONDS
...
+	int "Number of data transmissions to the server before shutdown"
+	range -1 2147483647
+	default -1
```
Comment (lemrey) [inline samples/cellular/udp/Kconfig:20]: I think the sample should terminate by default, so that we can run it in CI.

**Review by lemrey — CHANGES_REQUESTED** [review]

File: samples/cellular/udp/Kconfig (hunk 10 above)
Comment (MirkoCovizzi) [reply samples/cellular/udp/Kconfig:20]: @MechanicV1 From CI point of view should this sample terminate by default?

Comment (eivindj-nordic) [reply samples/cellular/udp/Kconfig:20]: CI can change the Kconfig value when building it. Though many other samples are terminating by default.

Comment (MechanicV1) [reply samples/cellular/udp/Kconfig:20]: Test automation will take care of setting the Kconfig ! @MirkoCovizzi

**Review by MechanicV1 — APPROVED** [review]

File: samples/cellular/udp/Kconfig (hunk 10 above)
Comment (lemrey) [reply samples/cellular/udp/Kconfig:20]: I think we should implement the same behavior as other sample. It's easy enough to change that for users if needed.
CI might want to override the value anyway to shorten the time it takes to complete the test.

**Review by lemrey — APPROVED** [review]

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 11]
```diff
@@ -316,6 +316,10 @@ Cellular samples
 
   * Added the configuration overlay file :file:`overlay-supl.conf` for building the sample with SUPL assistance support.
 
+* :ref:`udp` sample:
+
+  * Added the :kconfig:option:`CONFIG_UDP_DATA_UPLOAD_ITERATIONS` Kconfig option for configuring the number of data transmissions to the server.
```
Comment (divipillai) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:321]:
```suggestion
  * Added the :ref:`CONFIG_UDP_DATA_UPLOAD_ITERATIONS <CONFIG_UDP_DATA_UPLOAD_ITERATIONS>` Kconfig option for configuring the number of data transmissions to the server.
```

File: samples/cellular/udp/README.rst [hunk 12]
```diff
@@ -62,6 +62,12 @@ CONFIG_UDP_DATA_UPLOAD_SIZE_BYTES - UDP data upload size configuration
 CONFIG_UDP_DATA_UPLOAD_FREQUENCY_SECONDS - UDP data upload frequency configuration
    This configuration option sets the frequency with which the sample transmits data to the server.
 
+.. _CONFIG_UDP_DATA_UPLOAD_ITERATIONS:
+
+CONFIG_UDP_DATA_UPLOAD_ITERATIONS - UDP data upload iterations configuration
+   This configuration option sets the number of times the sample transmits data to the server before shutting down.
+   Set to -1 to transmit indefinitely.
```
Comment (divipillai) [inline samples/cellular/udp/README.rst:69]:
```suggestion
   Set to ``-1`` to transmit indefinitely.
```

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 11 above)
Comment (MirkoCovizzi) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:321]: @divipillai Is this the new standard syntax to use from now on?

Comment (divipillai) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:321]: This is the way to represent the application-specific Kconfig options. For example, for UDP sample, the ones that are defined here: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/cellular/udp/README.html#configuration. The Kconfig options defined in the Kconfig reference (library-specific) are provided with the ":kconfig:option:" syntax.

**Review by divipillai — APPROVED** [review]
